Sel branch new

voicelegion opened the pull request.

voicelegion committed:
QA

voicelegion committed:
QA: Added Selection of Dropdown value of Sell/buy/all, Selection of Dropdown value of Region, SEARCH.clicking,
sorting by price(price.click)

voicelegion committed:
Added ScreenShot, assert, Before class (and others)

voicelegion committed:
Added Wrapping for waiting (fluentwait) on element/page to load
Added Driver IF condition to determine which driver being used
Edited screenshot taker so it saves screenshots now in other direction and with time /date

voicelegion committed:
Added Min and Max value for price of product
Added creating of List containing all displayed elements

voicelegion committed:
working on 9th step of Ilja's task

voicelegion committed:
Working on 11th step of Ilja's task.
Added selecting of 3 web advertisments, saving text of each, and clicking of SHOW SELECTED button.
Now need to understand how to compare what we have with what should be

voicelegion committed:
Need to create assartions for 11th step

voicelegion committed:
Completed 11 th step, now need to refactor code. Changed advertisment picking from static numbers to random numbers

voicelegion committed:
Completed task, deleted unneeded variables (few ints)

krigan commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,35 +1,62 @@
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static junit.framework.TestCase.assertEquals;
+import static junit.framework.TestCase.assertTrue;
+import static org.junit.Assert.fail;
```
Please use asserts from one lib

krigan commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,35 +1,62 @@
 public class SSMission {
-    public static void main(String[] args) {
 
-        File file = new File("C:\\tools\\Firefox\\", "firefox" + ".exe");
```
"firefox" + ".exe" maybe would be better use "firefox.exe" ?

krigan commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,35 +1,62 @@
-        }
+    static WebDriver driver;
+    
+    public static String browser = "chrome";
```
make it constant

krigan commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,35 +1,62 @@
-        WebDriver driver = new FirefoxDriver();
+    @BeforeClass
+    public static void setup() {
+        if (browser.equals("chrome")) {
```
use swith/case instead of if/else in this case. or make browser strategies based on config file =)

krigan commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,35 +1,62 @@
 
+            File file = new File("C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe");
+            if (file.exists()) {
+                System.setProperty("webdriver.chrome.driver", "src\\test\\resources\\chromedriver.exe");
```
chromedriver is more suitable to store in src/main, not in src/test

krigan commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,35 +1,62 @@
+    public static void setup() {
+        if (browser.equals("chrome")) {
 
+            File file = new File("C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe");
```
extract constant with path

krigan commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,35 +1,62 @@
 
         driver.get("https://www.ss.lv");
+        driver.manage().window().maximize();
+        assertTrue(driver.findElement(By.xpath("//*[@href='/ru/']")).isDisplayed());
```
all locators extract to constants to avoid code duplication and make them reusable

krigan commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,35 +1,62 @@
+
+    @Test
+    public void testSS() throws InterruptedException {
 
```
implement page object pattern for code reusage

krigan commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -39,15 +66,107 @@ public static void main(String[] args) {
         wait.until(ExpectedConditions.elementToBeClickable(By.id("ptxt")));
-        driver.findElement(By.xpath("//*[@id='ptxt']")).sendKeys("Лаптоп");
+        driver.findElement(By.xpath("//*[@id='ptxt']")).sendKeys("Laptop");
+        Thread.sleep(1000);
```
avoid using Thread.sleep in tests

krigan commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,35 +1,62 @@
+        driver.manage().window().maximize();
+        assertTrue(driver.findElement(By.xpath("//*[@href='/ru/']")).isDisplayed());
         driver.findElement(By.xpath("//*[@href='/ru/']")).click();
         WebDriverWait wait = new WebDriverWait(driver, 5);
```
move WebDriverWait declaration to @Before method

krigan commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -39,15 +66,107 @@ public static void main(String[] args) {
+        wait.until(ExpectedConditions.elementToBeClickable(By.xpath("//*[text()=\"Цена\"][ancestor::*[@id=\"head_line\"]]")));
+        driver.findElement(By.xpath("//*[text()=\"Цена\"][ancestor::*[@id=\"head_line\"]]")).click();
+
+        Wait<WebDriver> stubbornWait = new FluentWait<WebDriver>(driver)
```
one more wait declaration? what for?

nikolajsokunevs reviewed (commented):
1. Page object!!! 
2. Update method 'setup()'. Use 'switch statement' instead of 'if'. Create 'enum' with all possible browsers
3. Remove ALL Thread.sleeps
4. Group actions into methods. Remove dublicates. Code should be readable
5. Test check nothing. Add Assestions.
6. Create constants for all static values, URL, WAITS, PATHS
7. Test name should reflex the test logic

krigan commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,35 +1,62 @@
 
-        System.setProperty("webdriver.gecko.driver", "src\\test\\resources\\geckodriver.exe");
-        WebDriver driver = new FirefoxDriver();
+    @BeforeClass
```
create base class for test and move driver initialization there

krigan commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,35 +1,62 @@
+        }
+    }
+
+    @Test
```
divide test code on methods, very hard to understand test purpose

krigan replied:
add loggers to test with steps description

voicelegion committed:
Started Creation of pageobjects

voicelegion committed:
Created HomePage, SearchPage, Electronics page, Developing SearchElectronics page.

voicelegion committed:
in SearchPage refactored locators from x path to Name / Id

voicelegion committed:
Finished SearchElectronicsPage ( changed name too )

voicelegion committed:
Finished ElectronicsSearchResultPag.

voicelegion committed:
Finished ElectronicsShowSelectedPage, and few minor changes in other Pages.

voicelegion committed:
Started to refactor test, using PageObjects. Ilja said to implement method "switchtoRu" - need to think about it

voicelegion committed:
Working on refactoring of Test. Refactored dropdowns logics in SearchElectronisPage, also added few enums (which is not best option, apperently).

voicelegion committed:
Last push, before refactoring to Page Factory. :/

voicelegion committed:
Refactoring Code adding FINDBY and other features of PageFactory.
Done changes in HomePage, Electronis Page, ElectronicsSHowSelectedPage, ElectronicsSearchResultPage and maybe SearcElectronicsPage

voicelegion committed:
Finished (?) integrating FINDBY in my PageObjects

voicelegion committed:
Commit, after facing problem when I am not able to get text of some ad, method from other class was not available there.
This is commit before i start fixing problem

voicelegion committed:
Finished replacing of code in TEST with PageObjects also using PageFactory

voicelegion committed:
Refactored final assertion, added sorting of 2 collections, containing selected ads texts and filtered ads texts

voicelegion committed:
Refactored final assertion, added sorting of 2 collections, containing selected ads texts and filtered ads texts

voicelegion committed:
Ilja edition

voicelegion committed:
After Ilja's changes reverted, ready to continue refactoring.

voicelegion committed:
OKAY SO! WIth Kolya's help I've created a method to create random number and method to create array of numbers using previously mentioned method. This happened in ElectronicsSearchResultPage and SSmission....

voicelegion committed:
DOne previoud task - created method (and used it) to create Last of ads texts, now developing method to select checkboxes.

pavelhrolovich commented on src/main/java/enums/Language.java:
```diff
@@ -0,0 +1,6 @@
+package enums;
```
Java has a convention how to name packages. Basically the naming designed to avoid namespace collisions. If you don't have domain name (like ctco.lv or swissre.com) you might use your email as a base, e.g. com.gmail.voicelegion and when add a package name.  Please note that naming conventions of packages are just that; conventions. You can name them whatever you want... The only enforced thing is that a . delineates a subdirectory in the filesystem where the compiler looks for the source.
Also, "enum" is a reserved java word, should not be used. All enums listed in the change are part of your domain. Better to move all to package like com.gmail.voicelegion.domain

pavelhrolovich commented on src/main/java/pageObjects/SearchElectronicsPage.java:
```diff
@@ -0,0 +1,58 @@
+                    break;
+                }
+                case SELL:{
+                    dropDown.selectByVisibleText("Продают");
```
If I will decide to test in a different language ? 
This better to isolate locale specific values, to move all to a property file.
See https://docs.oracle.com/javase/tutorial/i18n/resbundle/propfile.html

pavelhrolovich commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,53 +1,211 @@
+    public static void setup() {
+        if (browser.equals("chrome")) {
+
+            File file = new File("C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe");
```
First of all, any environment property must be moved outside of the code. You'll have to dance with a tambourine in case someone does not have chome in standard folder (for example, on linux server where CI will execute the test).
For intellij idea you can use -D flag in program arguments to pass environment specific properties. To retrieve any of required env property see https://docs.oracle.com/javase/tutorial/essential/environment/env.html

pavelhrolovich commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,53 +1,211 @@
+
+
+        wait.until(ExpectedConditions.urlContains("search/"));
+        assertEquals("https://www.ss.lv/ru/electronics/search/", driver.getCurrentUrl());
```
Same, locale specific resource.

pavelhrolovich commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,53 +1,211 @@
+        wait.until(ExpectedConditions.elementToBeClickable(searchElectronicsPage.searchTextBox));
+
+        searchElectronicsPage.searchTextBox.sendKeys("Laptop");
+        Thread.sleep(1000);
```
Really a need to use Thread.sleep? This force you to deal with exception handling. I see FluentWait class, it could be much better..

pavelhrolovich commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,53 +1,211 @@
-        }
+    static WebDriver driver;
+
+    public static String browser = "chrome";
```
This looks like you have two different test suites with the same scenario. One suite test on chrome, another on firefox. This might be done better with JUnit suite. You create two suite classes, run which one you need. You'll eliminate a property and an if statement, e.g.

```
@RunWith(Suite.class)
@SuiteClasses({SSMission.class})
public class TestSuiteChrome {°
    @Rule
    public TestRule screenshotTaker = new ScreenshotTaker((TakesScreenshot) driver);

    @BeforeClass
    public static void setUp() {
        System.out.println("setting up Chrome suite...");
 File file = new File("C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe");
    if (file.exists()) {
                System.setProperty("webdriver.chrome.driver", "src\\test\\resources\\chromedriver.exe");
                driver = new ChromeDriver();
            }
    }
}

```

Driver good to have somewhere in static context.

pavelhrolovich commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,53 +1,211 @@
+
+        ElectronicsPage electronicsPage = new ElectronicsPage(driver);
+
+        homePage.switchLangTo(Language.RU);
```
it's a test precondition, not a part of the test. Need to be placed to @ Before

pavelhrolovich commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,53 +1,211 @@
+        driver.get("https://www.ss.lv/lv");
+        driver.manage().window().maximize();
+
+        HomePage homePage = new HomePage(driver);
```
If you put driver to static context - you won't need to create beans in test. When with MockitoJUnitRunner this might be moved to injected field.

pavelhrolovich replied:
e.g. 
```

@Inject
private  HomePage homePage
```

pavelhrolovich commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,53 +1,211 @@
 
+    public static int randInt(int max) {
+
+        Random rand = new Random();
```
In Java 1.7 or later, the standard way to do this is as follows:
`return ThreadLocalRandom.current().nextInt(min, max + 1);`
 see http://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ThreadLocalRandom.html

pavelhrolovich commented on src/main/java/pageObjects/ElectronicsSearchResultPage.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    private int getRandInt(int max, int arrayOfRandomNumbers[]) {
+        Random rand = new Random();
```
use ThreadLocalRandom.current() instead

pavelhrolovich commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,53 +1,211 @@
+        ElectronicsShowSelectedPage electronicsShowSelectedPage = new ElectronicsShowSelectedPage(driver);
+
+        List<String> textsOfSelectedFilteredAds = new ArrayList<>();
+        textsOfSelectedFilteredAds.add(electronicsShowSelectedPage.getSelectedAdvText(electronicsShowSelectedPage.selectedAds.get(0)));
```
Why only 3 element, but not all ? Try to iterate via > electronicsShowSelectedPage.selectedAds and add all test to a list. Later, in case you have to find 4 elements - you will not need to do a code change.

pavelhrolovich replied:
btw, no need to gather all element to a list. You might check if ad is present in iterator.

pavelhrolovich commented on src/test/java/webdriverFramework/SSMission.java:
```diff
@@ -1,53 +1,211 @@
+
+        ElectronicsShowSelectedPage electronicsShowSelectedPage = new ElectronicsShowSelectedPage(driver);
+
+        List<String> textsOfSelectedFilteredAds = new ArrayList<>();
```
Initial size for array is a good practice. see http://stackoverflow.com/questions/15430247/why-start-an-arraylist-with-an-initial-capacity

pavelhrolovich commented on src/test/java/webdriverFramework/ScreenshotTaker.java:
```diff
@@ -0,0 +1,35 @@
+            throw new RuntimeException(e1);
+        }
+
+        System.err.println("taken " + screenshot);
```
Better to avoid System.err or System.out streams. It's too strict. You will not have any way to stream log to any other source or multiply sources. Take a look, for example, into Log4J or java logging. One of these will be preferable for you. When implement it.

pavelhrolovich reviewed (changes requested):
Propose to look / discuss proposed changes before we move forward.

voicelegion committed:
Trying to replace tests code with many lines to methods, to make it shorter. Done a lot of changes, added WaitDriver to each constructor so i can remove waits from test code

voicelegion committed:
Improved code to make WAIT work in each page object constructor, implemented fillFieldsAndSubmit method for searchElectronisPage, and used hashmap test.

voicelegion committed:
Almost finished refactoring of my code, made it much much shorted. Will not continue now. New task from Ilja is to read about I/O.

voicelegion committed:
Practiced a little with IO, used config.properties, hard stuff :/

voicelegion committed:
Practiced a little with IO, used config.properties, hard stuff :/ vers2

voicelegion committed:
Practiced a little with IO, used config.properties, hard stuff :/ vers2

voicelegion committed:
Refactored code: - deleted commented code lines,
- reworked screenshotTaker
-added allure reports with screenshots
-driver initialization mechanism - DriverFactory
-edited structure of project. Ready to merge with main
